larray/core/session.py

```diff
@@ -1561,14 +1533,55 @@ class ConstrainedSession(Session):
     Content of file 'main.py'
 
     >>> run_model('projection_2020_2070', first_proj_year=2020)
+    dumping FIRST_OBS_YEAR ... done
+    dumping LAST_PROJ_YEAR ... done
+    dumping AGE ... done
+    dumping GENDER ... done
+    dumping G_CHILDREN ... done
+    dumping G_ADULTS ... done
     dumping variant_name ... done
     dumping first_proj_year ... done
-    dumping g_obs_years ... done
-    dumping g_proj_years ... done
-    dumping population ... done
-    dumping births ... done
-    dumping deaths ... done
+    dumping obs_years ... done
+    dumping proj_years ... done
+    dumping population_obs ... done
+    dumping population_proj ... done
+    dumping mortality_rate ... done
     """
+    def __init__(self, *args, **kwargs):
+        _cls_attrs = {key: value for key, value in vars(self.__class__).items() if not key.startswith('_')}
+        object.__setattr__(self, '_cls_attrs', _cls_attrs)
+
+        constant_cls_attrs = {key: value for key, value in _cls_attrs.items()
+                              if not isinstance(value, (type, ArrayDef))}
+
+        if len(args) == 1:
+            # Note: using the two lines below we may loose the order in which variables have been declared
+            # but can we avoid it?
+            Session.__init__(self, *args, **kwargs)
+            self.add(**constant_cls_attrs)
+        else:
+            # Note: same as above
+            kwargs.update(constant_cls_attrs)
+            Session.__init__(self, *args, **kwargs)
+
+    def load(self, fname, names=None, engine='auto', display=False, **kwargs):
+        super().load(fname, names, engine, display, **kwargs)
+        for key in set(self._cls_attrs.keys()) - set(self.keys()):
+            warnings.warn(f"The variable {key} is declared in the {self.__class__.__name__} "
+                          f"class definition but has not found in file {fname}.")
+        for key in set(self.keys()) - set(self._cls_attrs.keys()):
+            warnings.warn(f"The variable {key} has been found in file {fname} but is not declared "
+                          f"in the  {self.__class__.__name__} class definition.")
+
+    def save(self, fname, names=None, engine='auto', overwrite=True, display=False, **kwargs):
+        for key in set(self._cls_attrs.keys()) - set(self.keys()):
```

Comment:
There should **not** be such checks and warnings during save! Those should have happened earlier (in `__setitem__` and `__setattr__` -- which is already the case AFAICT)!